### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by joernhees** [openlink/virtuoso-opensource#645]
is there anything that i can do to make virtuoso obey the memory limits specified in its ini?

the following server (256GB ram) runs a virtuoso instance that is set up as recommended for 64 GB, but as you can see it nibbles 168 GB of residual memory...

![screen shot 2017-04-23 at 19 08 09](https://cloud.githubusercontent.com/assets/464192/25315616/42f61b88-2858-11e7-9e63-1efca8a4aeba.png)

The server is set up according to https://joernhees.de/blog/2015/11/23/setting-up-a-linked-data-mirror-from-rdf-dumps-dbpedia-2015-04-freebase-wikidata-linkedgeodata-with-virtuoso-7-2-1-and-docker-optional/ and contains ~ 8 G triples...

In its current state i can't run the (read only) DB on nodes with less than 64 GB for more than 2 hours (load by my [graph pattern learner algorithm](https://github.com/RDFLib/graph-pattern-learner), max 16 queries in parallel, each with a timeout of max 2s, ~500K queries in total before death by Kernel OOM).

**Comment by joernhees** [comment]
actually the issue doesn't seem to be limited to our main dump, but also a comparably small one of just the dbpedia + wikilinks (~ 600 M triples)... that is configured to run with 32 GB, if i run that on a 64 GB server it uses > 50 GB within ~3 hours as well, then crashes due to out of memory...

doesn't seem to be a new issue, the more i search... #128 #356 :-/
any tips how to mitigate?

**Comment by joernhees** [comment]
```
virtuoso-t --version
Virtuoso Open Source Edition (Column Store) (multi threaded)
Version 7.2.4.2.3217-pthreads as of Apr  9 2017
Compiled for Linux (x86_64-unknown-linux-gnu)
Copyright (C) 1998-2016 OpenLink Software
```

i already tried playing around with the amount of connections, threads, cleanup, etc...

unlike the other issues i don't think this can be pinned to individual queries... i see steady slow growth while the endpoint is under load, that's why i assume it's a memory leak.

```
[Database]
DatabaseFile                    = virtuoso-opensource-7/db/virtuoso.db
ErrorLogFile                    = virtuoso-opensource-7/db/virtuoso.log
LockFile                        = virtuoso-opensource-7/db/virtuoso.lck
TransactionFile                 = virtuoso-opensource-7/db/virtuoso.trx
xa_persistent_file              = virtuoso-opensource-7/db/virtuoso.pxa
ErrorLogLevel                   = 7
FileExtend                      = 200
MaxCheckpointRemap              = 312500
Striping                        = 0
TempStorage                     = TempDatabase


[TempDatabase]
DatabaseFile                    = virtuoso-opensource-7/db/virtuoso-temp.db
TransactionFile                 = virtuoso-opensource-7/db/virtuoso-temp.trx
MaxCheckpointRemap              = 2000
Striping                        = 0


[Parameters]
ServerPort                      = 1111
LiteMode                        = 0
DisableUnixSocket               = 1
DisableTcpSocket                = 0
MaxClientConnections            = 10
CheckpointInterval              = 60
O_DIRECT                        = 0
CaseMode                        = 2
MaxStaticCursorRows             = 5000
CheckpointAuditTrail            = 0
AllowOSCalls                    = 0
SchedulerInterval               = 10
DirsAllowed                     = .
ThreadCleanupInterval           = 10
ThreadThreshold                 = 10
ResourcesCleanupInterval        = 0
FreeTextBatchSize               = 100000
SingleCPU                       = 0
VADInstallDir                   = /home/hees/opt/virtuoso-opensource/share/virtuoso/vad/
PrefixResultNames               = 0
RdfFreeTextRulesSize            = 100
IndexTreeMaps                   = 256
MaxMemPoolSize                  = 200000000
PrefixResultNames               = 0
MacSpotlight                    = 0
IndexTreeMaps                   = 64
MaxQueryMem                     = 1G            ; memory allocated to query processor
VectorSize                      = 1000          ; initial parallel query vector (array of query operations) size
MaxVectorSize                   = 1000000       ; query vector size threshold.
AdjustVectorSize                = 0
ThreadsPerQuery                 = 4
AsyncQueueMaxThreads            = 10
NumberOfBuffers          = 2720000
MaxDirtyBuffers          = 2000000

[HTTPServer]
ServerPort                      = 8890
ServerRoot                      = /home/hees/opt/virtuoso-opensource/var/lib/virtuoso/vsp
MaxClientConnections            = 32
DavRoot                         = DAV
EnabledDavVSP                   = 0
HTTPProxyEnabled                = 0
TempASPXDir                     = 0
DefaultMailServer               = localhost:25
ServerThreads                   = 10
MaxKeepAlives                   = 10
KeepAliveTimeout                = 10
MaxCachedProxyConnections       = 10
ProxyConnectionCacheTimeout     = 15
HTTPThreadSize                  = 280000
HttpPrintWarningsInOutput       = 0
Charset                         = UTF-8
MaintenancePage                 = atomic.html
EnabledGzipContent              = 0

[SPARQL]
ResultSetMaxRows                = 100000
MaxQueryCostEstimationTime      = 600   ; in seconds
MaxQueryExecutionTime           = 1200  ; in seconds
DefaultQuery                    = select ?g count(*) as ?c where { graph ?g { ?s ?p ?o } } order by desc(?c)
DeferInferenceRulesInit         = 0  ; controls inference rules loading
```

**Comment by HughWilliams** [comment]
Looking at your Virtuoso version `7.2.4.2.3217` this looks like a build from the git stable/7 branch that was released on Apr 2016 ?  There have been a number of memory leak fixes in the the Virtuoso git develop/7 branch (https://github.com/openlink/virtuoso-opensource), which I would recommend a build be performed against and tested with ...

BTW, separately looking at your INI file I not in the "[Parameters]" section you have "MaxClientConnections  = 10", I would recommend setting it to something like "100" as this controls the number of SQL server threads that can be allocated by the Virtuoso server, some of which are used for SQL client connections and also for other internal threads required by the Virtuoso database. And as these are allocated on demand (unlike the HTTP threads which are preallocated on server startup for the HTTP thread pool) there is no harm in setting this to a value of 100 as they will only be allocated when required ...

**Comment by joernhees** [comment]
thanks for the quick reply and the tip wrt `MaxClientConnections`.

i thought the version i'm using is already the latest stable version?!? is there a more recent stable one?

**Comment by HughWilliams** [comment]
As said previous you are running the latest stable/7 which is from April last year and I am asking you to test with the current develop/7 which has all the latest fixes and is being prep'ed to become then new stable/7 release, to see if this resolves your problems ...

**Comment by LLTommy** [comment]
@joernhees Did you try this? Any experiences that you can share with the develop/7 version and memory leaks? Did it help?
